src/main/scala/com/gu/support/workers/model/CheckoutFailureReasons.scala

```diff
@@ -0,0 +1,92 @@
+package com.gu.support.workers.model
+
+object CheckoutFailureReasons {
+
+  val all = List(
+    PaymentMethodDetailsIncorrect,
+    PaymentMethodTemporarilyDeclined,
+    PaymentMethodUnacceptable,
+    PaymentProviderUnavailable,
+    PaymentRecentlyTaken,
+    Unknown
+  )
+
+  def fromString(string: String): Option[CheckoutFailureReason] = all.find(_.asString == string)
+
+  sealed trait CheckoutFailureReason {
+    def asString: String
+  }
+
+  case object PaymentMethodDetailsIncorrect extends CheckoutFailureReason {
+    override def asString: String = "payment_details_incorrect"
+  }
+
+  case object PaymentMethodTemporarilyDeclined extends CheckoutFailureReason {
+    override def asString: String = "payment_method_temporarily_declined"
+  }
+
+  case object PaymentMethodUnacceptable extends CheckoutFailureReason {
+    override def asString: String = "payment_method_unacceptable"
+  }
+
+  case object PaymentProviderUnavailable extends CheckoutFailureReason {
+    override def asString: String = "payment_provider_unavailable"
+  }
+
+  case object PaymentRecentlyTaken extends CheckoutFailureReason {
+    override def asString: String = "payment_recently_taken"
+  }
+
+  case object Unknown extends CheckoutFailureReason {
+    override def asString: String = "unknown"
+  }
+
+  // https://stripe.com/docs/declines/codes
+  def convertStripeDeclineCode(declineCode: String): Option[CheckoutFailureReason] = declineCode match {
+    case "approve_with_id" => Some(PaymentMethodTemporarilyDeclined)
+    case "call_issuer" => Some(PaymentMethodUnacceptable)
+    case "card_not_supported" => Some(PaymentMethodUnacceptable)
+    case "card_velocity_exceeded" => Some(PaymentMethodUnacceptable)
+    case "currency_not_supported" => Some(PaymentMethodUnacceptable)
+    case "transaction_not_allowed" => Some(PaymentMethodUnacceptable)
+    case "do_not_honor" => Some(PaymentMethodUnacceptable)
+    case "do_not_try_again" => Some(PaymentMethodUnacceptable)
+    case "duplicate_transaction" => Some(PaymentRecentlyTaken)
+    case "expired_card" => Some(PaymentMethodUnacceptable)
+    case "fraudulent" => Some(PaymentMethodUnacceptable)
+    case "generic_decline" => Some(PaymentMethodUnacceptable)
+    case "incorrect_number" => Some(PaymentMethodDetailsIncorrect)
+    case "incorrect_cvc" => Some(PaymentMethodDetailsIncorrect)
+    case "incorrect_pin" => Some(PaymentMethodDetailsIncorrect)
+    case "incorrect_zip" => Some(PaymentMethodDetailsIncorrect)
+    case "insufficient_funds" => Some(PaymentMethodUnacceptable)
```

Comment:
Might be worth including an extra case for this? It's the sort of detail a user would be interested in.

Reply:
Yes, I think that's a good idea - it's also quite a common case, so it seems worth having a specific message client-side 👍